walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,76 @@
 
 package com.google.sps;
 
+
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    //Checks if we have a valid meeting duration and attendees
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    } else if(request.getAttendees().isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    } else {
+        Collection<TimeRange> validTimes = new ArrayList<TimeRange>();
+        Event[] attendeeEvents = getAttendeeEvents(events, request.getAttendees());
+        
+        int start = TimeRange.START_OF_DAY;
+        int i = 0;
+
+        while (i < attendeeEvents.length) {
+            //Create a new time range the start to the beginning of next event
+            TimeRange valid = TimeRange.fromStartEnd(start, attendeeEvents[i].getWhen().start(), false);
+            if (valid.duration() >= request.getDuration()) {
+                validTimes.add(valid);
+            }
+                
+            if ((i+1) < attendeeEvents.length && attendeeEvents[i].getWhen().overlaps(attendeeEvents[i+1].getWhen())) {
+                start = (attendeeEvents[i].getWhen().end() > attendeeEvents[i+1].getWhen().end() ? attendeeEvents[i].getWhen().end() : attendeeEvents[i + 1].getWhen().end());
+                i++;
```

Comment:
This only checks the next event overlaps, but you may have more than 1 event overlapping (or even multiple events completely contained) if you have a single event that is really long, and a bunch of shorter events that all start after the long event.

I think you could do this check before checking a valid slot with enough duration exists by adding a few checks:

// If the end time of this event is before the next available start time `start`, continue to the next event.
if (attendeeEvents[i].getWhen().end() < start) {
    continue;
}

// If the current event contains the next available start time `start`, update `start` to the end of this 
// event and continue to the next event.
if(attendeeEvents[i].getWhen().contains(start)) {
   start = attendeeEvents[i].getWhen().end();
   continue;
}

//Check valid range and add to validTimes if duration is long enough
...